nhrpd/nhrp_shortcut.c

```diff
@@ -262,39 +291,58 @@ static void nhrp_shortcut_recv_resolution_rep(struct nhrp_reqid *reqid,
 	}
 
 	debugf(NHRP_DEBUG_COMMON,
-	       "Shortcut: %s is at proto %s cie-nbma %s nat-nbma %s cie-holdtime %d",
+	       "Shortcut: %s is at proto %s dst_proto %s cie-nbma %s nat-nbma "
+	       "%s cie-holdtime %d",
 	       prefix2str(&prefix, bufp, sizeof bufp),
-	       sockunion2str(proto, buf[0], sizeof buf[0]),
-	       sockunion2str(&cie_nbma, buf[1], sizeof buf[1]),
-	       sockunion2str(&nat_nbma, buf[2], sizeof buf[2]),
+	       sockunion2str(proto, buf[0], sizeof(buf[0])),
+	       sockunion2str(&pp->dst_proto, buf[1], sizeof(buf[1])),
+	       sockunion2str(&cie_nbma, buf[2], sizeof(buf[2])),
+	       sockunion2str(&nat_nbma, buf[3], sizeof(buf[3])),
 	       htons(cie->holding_time));
 
 	/* Update cache entry for the protocol to nbma binding */
-	if (sockunion_family(&nat_nbma) != AF_UNSPEC) {
+	if (sockunion_family(&nat_nbma) != AF_UNSPEC)
 		nbma = &nat_nbma;
-		nbma_natoa = &cie_nbma;
-	} else {
+	else
 		nbma = &cie_nbma;
-		nbma_natoa = NULL;
-	}
+
 	if (sockunion_family(nbma)) {
 		c = nhrp_cache_get(pp->ifp, proto, 1);
 		if (c) {
-			nhrp_cache_update_binding(c, NHRP_CACHE_CACHED,
+			debugf(NHRP_DEBUG_COMMON,
+			       "Shortcut: cache found, update binding");
+			nhrp_cache_update_binding(c, NHRP_CACHE_DYNAMIC,
 						  holding_time,
 						  nhrp_peer_get(pp->ifp, nbma),
-						  htons(cie->mtu), nbma_natoa);
+						  htons(cie->mtu), nbma);
+		} else {
+			debugf(NHRP_DEBUG_COMMON,
+			       "Shortcut: no cache for nbma %s", buf[2]);
 		}
 	}
 
 	/* Update shortcut entry for subnet to protocol gw binding */
-	if (c && !sockunion_same(proto, &pp->dst_proto)) {
+	if (c) {
 		ps = nhrp_shortcut_get(&prefix);
 		if (ps) {
 			ps->addr = s->addr;
-			nhrp_shortcut_update_binding(ps, NHRP_CACHE_CACHED, c,
+			debugf(NHRP_DEBUG_COMMON,
+			       "Shortcut: "
+			       "calling nhrp_shortcut_update_binding");
+			nhrp_shortcut_update_binding(ps, NHRP_CACHE_DYNAMIC, c,
```

Comment:
RFC2332 mention of 'dynamics' is completely irrelated matter. `NHRP_CACHE_` are nhrpd internal states. This breaks the internal logic of code elsewhere. This should be reverted.